File: rust/cudf-sys/cpp/hashing.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#include "cudf-sys/src/hashing.rs.h"
#include "cudf-sys/src/lib.rs.h"
#include "cudf-sys/helpers.hpp"

#include <cudf/hashing.hpp>

namespace cudf_sys {

std::unique_ptr<Column> hash_murmur3(Table const& tbl, uint32_t seed, std::size_t stream) {
  return COL(cudf::hashing::murmurhash3_x86_32(tbl.cached_view(), seed, S(stream)));
}

std::unique_ptr<Column> hash_xxhash64(Table const& tbl, uint64_t seed, std::size_t stream) {
  return COL(cudf::hashing::xxhash_64(tbl.cached_view(), seed, S(stream)));
}

std::unique_ptr<Column> hash_md5(Table const& tbl, std::size_t stream) {
  return COL(cudf::hashing::md5(tbl.cached_view(), S(stream)));
}

std::unique_ptr<Column> hash_sha256(Table const& tbl, std::size_t stream) {
  return COL(cudf::hashing::sha256(tbl.cached_view(), S(stream)));
}

std::unique_ptr<Table> hash_murmurhash3_x64_128(Table const& tbl, uint64_t seed, std::size_t stream) {
  return TBL(cudf::hashing::murmurhash3_x64_128(tbl.cached_view(), seed, S(stream)));
}

std::unique_ptr<Column> hash_sha1(Table const& tbl, std::size_t stream) {
  return COL(cudf::hashing::sha1(tbl.cached_view(), S(stream)));
}

std::unique_ptr<Column> hash_xxhash_32(Table const& tbl, uint32_t seed, std::size_t stream) {
  return COL(cudf::hashing::xxhash_32(tbl.cached_view(), seed, S(stream)));
}

std::unique_ptr<Column> hash_sha224(Table const& tbl, std::size_t stream) {
  return COL(cudf::hashing::sha224(tbl.cached_view(), S(stream)));
}

std::unique_ptr<Column> hash_sha384(Table const& tbl, std::size_t stream) {
  return COL(cudf::hashing::sha384(tbl.cached_view(), S(stream)));
}

std::unique_ptr<Column> hash_sha512(Table const& tbl, std::size_t stream) {
  return COL(cudf::hashing::sha512(tbl.cached_view(), S(stream)));
}

std::unique_ptr<Column> hash_murmurhash3_x86_32(Table const& tbl, uint32_t seed, std::size_t stream) {
  return COL(cudf::hashing::murmurhash3_x86_32(tbl.cached_view(), seed, S(stream)));
}

}  // namespace cudf_sys

File: rust/cudf-sys/src/hashing.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#[cxx::bridge(namespace = "cudf_sys")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cudf-sys/lib.hpp");

        // Shared types from the core bridge.
        type Column = crate::ffi::Column;
        type Table = crate::ffi::Table;

        // -- Hashing --

        /// Computes `MurmurHash3` 32-bit hash of each row.
        fn hash_murmur3(tbl: &Table, seed: u32, stream: usize) -> UniquePtr<Column>;

        /// Computes `XXHash64` hash of each row.
        fn hash_xxhash64(tbl: &Table, seed: u64, stream: usize) -> UniquePtr<Column>;

        /// Computes MD5 hash of each row (returns string column).
        fn hash_md5(tbl: &Table, stream: usize) -> UniquePtr<Column>;

        /// Computes SHA-256 hash of each row (returns string column).
        fn hash_sha256(tbl: &Table, stream: usize) -> UniquePtr<Column>;

        /// `MurmurHash3` 128-bit hash (returns table of two UINT64 columns).
        fn hash_murmurhash3_x64_128(
            tbl: &Table,
            seed: u64,
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// SHA-1 hash of each row (returns string column).
        fn hash_sha1(tbl: &Table, stream: usize) -> Result<UniquePtr<Column>>;

        /// `XXHash` 32-bit hash of each row.
        fn hash_xxhash_32(tbl: &Table, seed: u32, stream: usize) -> Result<UniquePtr<Column>>;

        /// SHA-224 hash of each row (returns string column).
        fn hash_sha224(tbl: &Table, stream: usize) -> Result<UniquePtr<Column>>;

        /// SHA-384 hash of each row (returns string column).
        fn hash_sha384(tbl: &Table, stream: usize) -> Result<UniquePtr<Column>>;

        /// SHA-512 hash of each row (returns string column).
        fn hash_sha512(tbl: &Table, stream: usize) -> Result<UniquePtr<Column>>;

        /// `MurmurHash3` x86 32-bit hash of each row (returns UINT32 column).
        fn hash_murmurhash3_x86_32(
            tbl: &Table,
            seed: u32,
            stream: usize,
        ) -> Result<UniquePtr<Column>>;
    }
}
